src/Components/Components/src/NavigationManager.cs

```diff
@@ -30,6 +30,29 @@ public abstract class NavigationManager
 
         private EventHandler<LocationChangedEventArgs>? _locationChanged;
 
+        /// <summary>
+        /// An event that fires when the navigation location is about to change
+        /// </summary>
+        public event EventHandler<LocationChangingEventArgs> LocationChanging
```

Comment:
Feedback from API review: we discussed the risks of implementing the `LocationChanging` API as an event handler, especially with regard to async scenarios. For example, if a user wants to send an HTTP request to determine whether the users content has been saved before navigating.

Instead of event, we discussed maintaining an internal list of of navigation listeners.

```csharp
private List<Func<LocationChangingContext, ValueTask>> _locationChangingListeners ;
```

The `LocationChangingContext` property above contains the same properties as the `LocationChangingEventArgs` but is renamed since we're no longer implementing the event handler pattern here. The listener the user provides will return a `ValueTask` that can be awaited.

Since the `_locationChangingListeners` property is internal, we discussed implementing a `RegisterLocationChangingListener` method on the `NavigationManager` that implements the following API.

```csharp
public void RegisterLocationChangingListener(Func<LocationChangingContext, ValueTask> listener)
```

When invoked, this method will add the provided `listener` to the internally maintained list.

When it comes to using the registered ` _locationChangingListeners`, we'll await each of the listeners in the handle location changing API. If any one of the listeners returns `false`, we do not await the remaining listeners.

Reply:
Good catch, even marking an event handler as async and awaiting a call would not have had the intended effect, due to Blazor's SynchronizationContext.

I do have some additional remarks about the above approach. I preface my remarks with my expectation that the main use cases for the LocationChanging 'event' is the ability of a user to either cancel the current navigation or do some emergency housekeeping just before navigation would occur.

I would want to keep the ability to add and remove listeners so that users can add listeners on a per component base level. The above proposal hints towards a global registration of handlers. (I base my observation on the fact that you are using a register function, and ASPNET Core doesn't use the Register/Unregister pair, and that you are registering a Func<>)

My second remark is about:
> If any one of the listeners returns false, we do not await the remaining listeners.

~~I think this might introduce some confusion and would invalidate some use cases (In particular the use case where a user would do some emergency housekeeping (on LocationChanging).
I would either always call the registered listeners, or add an additional boolean to indicate that this listener should always be called even when another listener canceled.~~

_After some more personal deliberation, I would agree with your proposal._ 


My changed proposal would be something like.

```cs
public class LocationChangingContext 
{
  public string Location { get; }
  public bool IsNavigationIntercepted { get; }
}
```

Adding an Interface
```cs
public interface ILocationChanging
{
  ValueTask<bool> OnLocationChanging(LocationChangingContext context);
}
```

Add some functions to ```NavigationManager```
```cs
  public void AddLocationChangingListener(ILocationChanging locationChanging);
  public bool RemoveLocationChangingListener(ILocationChanging locationChanging);
```

I prefer an interface instead of
```cs
  Func<LocationChangingContext, ValueTask<bool>>
```
Because it can be easily removed. and users cannot simply use something like
```cs
  Add( async (ctx) =>await ShouldWeCancel() );
```
which would prevent users from removing the above function, and could introduce memory leaks.

Reply:
I am going to be a little bold, and mark this as resolved, this does not by any means mean that the discussion is over, but I have committed code that implements this API change roughly as I described above. There are some questions I have, but I will add that to another comment, And feel free to criticize my implementation.